### PR TITLE
GH-118447: Fix handling of unreadable symlinks in `os.path.realpath()`

### PR DESCRIPTION
**Pull request by barneygale**
* Issue: gh-118447

### CODE REVIEW

**Comment by nineteendo** [comment]
Do not merge this, this will get tricked by `secret-symlink/..`: https://github.com/python/cpython/issues/118447#issuecomment-2089616796.
But it is a bug in coreutils, so I'll create an issue there.

**Review by nineteendo — CHANGES_REQUESTED** [review]
If not resolving unreadable symlinks is acceptable, you still need to raise an error for `link/..`.

**Comment by barneygale** [comment]
> If not resolving unreadable symlinks is acceptable, you still need to raise an error for `link/..`.

`realpath()` shouldn't raise `OSError` in non-strict mode.

**Comment by nineteendo** [comment]
@serhiy-storchaka, do you agree with me? We have to choose between an error and incorrect behaviour for `unreadable_link/..`

**Comment by nineteendo** [comment]
OK, technically you could also keep `unreadable_link/..`, but that's going to be a lot of code to fix this edge case.

File: Lib/posixpath.py
Comment (serhiy-storchaka) [inline Lib/posixpath.py]: Why not keep `os.stat(newpath)` here?

File: Lib/test/test_posixpath.py
Comment (serhiy-storchaka) [inline Lib/test/test_posixpath.py:664]: Is it needed? How does the test work on Windows?

File: Lib/test/test_posixpath.py
Comment (serhiy-storchaka) [inline Lib/test/test_posixpath.py:670]: Please test also cases where non-readable symlink is followed by  other path components (including `..`), e.g. `badlink/foo`, `badlink/../foo`, `badlink/foo/..`, etc.

If the following `..` removes `badlink`, try to test something that involves a loop containing `badlink`, if this is possible. Your code does not set `seen[newpath] = None` before calling `readlink()`, I wonder what effect does it have.

File: Lib/posixpath.py
Comment (barneygale) [reply Lib/posixpath.py]: For consistency with the way we raise `NotADirectoryError` in #118290, and to save the small cost of a `stat()`

File: Lib/test/test_posixpath.py
Comment (barneygale) [reply Lib/test/test_posixpath.py:664]: It won't work on Windows because `ABSTFN` is an absolute path with backslash separators. These decorators are also used for other tests in this suite.

File: Lib/test/test_posixpath.py
Comment (nineteendo) [reply Lib/test/test_posixpath.py:670]: This is the behaviour we would be emulating:

```zsh
wannes@Stefans-iMac ~ % ln -s . src        
wannes@Stefans-iMac ~ % grealpath -e src/..
/Users
wannes@Stefans-iMac ~ % chmod -h 000 src
wannes@Stefans-iMac ~ % grealpath -e src/..
/Users/wannes
```

Is that not a problem, unlike the behaviour of `abspath()` compared to `realpath()`?

Comment (nineteendo) [reply Lib/test/test_posixpath.py:670]:
```suggestion
            self.assertEqual(realpath(ABSTFN), ABSTFN)
            self.assertEqual(realpath(ABSTFN + '/foo'), ABSTFN + '/foo')
            self.assertEqual(realpath(ABSTFN + '/../foo'), dirname(ABSTFN) + '/foo')
            self.assertEqual(realpath(ABSTFN + '/foo/..'), ABSTFN)
```

**Review by nineteendo — COMMENTED** [review]
OK, you're right, it shouldn't raise an error. That's what `strict` is for.

**Comment by serhiy-storchaka** [comment]
I think that it is reasonable to expect that `realpath('dir/badlink/rest')` returns the same result when `badlink` does not exist, is a file, is in non-readable directory or is non-readable itself, for any `rest`. It's error prone, but that's how it was defined.

**Review by nineteendo — APPROVED** [review]
Barney, would you like to continue working on this? You have my approval if it means anything.

File: Lib/posixpath.py
Comment (barneygale) [reply Lib/posixpath.py]: Well, there's no point in being consistent with something that hasn't landed yet, so I'll switch back to `os.stat(newpath)`. Thanks for your suggestion.

**Comment by barneygale** [comment]
Thanks for the reviews and suggestions, both!

**Comment by barneygale** [comment]
If no one objects, I'll skip backporting to 3.12 as this is a fairly minor bug and I can't be bothered to fix the conflicts :)

**Pull request merged**
